File: src/ui-shared/format.ts
// Presentation helpers with no framework in them, so every JavaScript UI formats a timestamp the
// same way. Two dashboards rendering the same instant differently is a small bug that is very
// confusing to hit.

/** UTC, to the second, always — job times are UTC and a local rendering invites misreading. */
export function formatTime(value: string | null | undefined): string {
  if (!value) return '—'
  const date = new Date(value)
  return Number.isNaN(date.valueOf())
    ? '—'
    : date.toISOString().replace('T', ' ').slice(0, 19) + 'Z'
}

export function relativeToNow(value: string): string {
  const delta = new Date(value).getTime() - Date.now()
  const abs = Math.abs(delta)
  const minute = 60_000
  const hour = 3_600_000
  const day = 86_400_000

  let text: string
  if (abs < minute) text = `${Math.round(abs / 1000)}s`
  else if (abs < hour) text = `${Math.round(abs / minute)}m`
  else if (abs < day) text = `${Math.round(abs / hour)}h`
  else text = `${Math.round(abs / day)}d`

  return delta >= 0 ? `in ${text}` : `${text} ago`
}

export function isOverdue(nextFireTime: string): boolean {
  return new Date(nextFireTime).getTime() < Date.now()
}

/** "overdue by 3h" or "in 20m" — the schedule view's one piece of derived text. */
export function dueText(nextFireTime: string): string {
  const relative = relativeToNow(nextFireTime)
  return isOverdue(nextFireTime) ? `overdue by ${relative.replace(' ago', '')}` : relative
}

export function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error)
}

File: src/Millrace.Dashboard.Ui.Blazor/ui/Format.cs
using System.Globalization;

namespace Millrace.Dashboard.Ui.Blazor.App;

/// <summary>
/// Presentation helpers, matching <c>ui-shared/format.ts</c> exactly.
/// </summary>
/// <remarks>
/// A deliberate second implementation rather than a shared one. The TypeScript UIs share these
/// because they share a language; this is the same rules in C#, and the rules are the contract —
/// three dashboards rendering the same instant differently is a small bug that is very confusing to
/// hit. `FormatTimeTests` in the dashboard test project pins the two against each other.
/// </remarks>
public static class Format
{
    /// <summary>UTC, to the second, always — job times are UTC and a local rendering invites misreading.</summary>
    public static string Time(DateTimeOffset? value) => value is null
        ? "—"
        : value.Value.UtcDateTime.ToString("yyyy-MM-dd HH:mm:ss'Z'", CultureInfo.InvariantCulture);

    /// <summary>"in 20m" or "3h ago", to one unit.</summary>
    public static string RelativeToNow(DateTimeOffset value, DateTimeOffset now)
    {
        var delta = value - now;
        var abs = delta.Duration();

        var text = abs.TotalMinutes < 1 ? $"{Math.Round(abs.TotalSeconds):N0}s"
            : abs.TotalHours < 1 ? $"{Math.Round(abs.TotalMinutes):N0}m"
            : abs.TotalDays < 1 ? $"{Math.Round(abs.TotalHours):N0}h"
            : $"{Math.Round(abs.TotalDays):N0}d";

        return delta >= TimeSpan.Zero ? $"in {text}" : $"{text} ago";
    }

    /// <summary>Whether a schedule's next fire time has already passed.</summary>
    public static bool IsOverdue(DateTimeOffset nextFireTime, DateTimeOffset now) => nextFireTime < now;

    /// <summary>"overdue by 3h" or "in 20m" — the schedule view's one piece of derived text.</summary>
    public static string DueText(DateTimeOffset nextFireTime, DateTimeOffset now)
    {
        var relative = RelativeToNow(nextFireTime, now);
        return IsOverdue(nextFireTime, now)
            ? $"overdue by {relative.Replace(" ago", string.Empty, StringComparison.Ordinal)}"
            : relative;
    }

    /// <summary>What to show an operator when a request failed.</summary>
    public static string ErrorMessage(Exception error) => error.Message;
}
